### PR TITLE
Bug/7596 rc0 addon.ocr.tesseract is faulty

### PR DESCRIPTION
**Pull request by DanielWilczek**

### CODE REVIEW

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 1]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
         public OcrOfflineFindCommand(AbstractScripter scripter) : base(scripter)
         {
         }
 
-        double imgRescaleRatio = 4.0;
         public void Execute(Arguments arguments)
         {
-            var rectangle = !arguments.Relative.Value ? arguments.Area.Value : arguments.Area.Value.ToAbsoluteCoordinates();
+            if (!arguments.Area.Value.IsValidRectangle())
+                throw new ArgumentException("Argument Area is not a valid rectangle");
```
Comment (danieladamkaiser) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: Just a note:
Do we really want to remove rectangles with negative width/height? They are valid Rectangle struct and allow user to "draw" rectangle from different point

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 2]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
+            var rectangle = arguments.Area.Value;
+            if (arguments.Relative.Value)
+            {
+                var foregroundWindowRect = new RobotWin32.Rect();
+                RobotWin32.GetWindowRectangle(RobotWin32.GetForegroundWindow(), ref foregroundWindowRect);
+                rectangle = new Rectangle(rectangle.X + foregroundWindowRect.Left,
+                    rectangle.Y + foregroundWindowRect.Top,
+                    foregroundWindowRect.Right - foregroundWindowRect.Left,
+                    foregroundWindowRect.Bottom - foregroundWindowRect.Top);
```
Comment (danieladamkaiser) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: Why we have these 2 lines here? Width and Height should be from rectangle provided by user

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 3]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
 
             [Argument(Tooltip = "The language which should be considered trying to recognize text")]
             public TextStructure Language { get; set; } = new TextStructure("eng");
+
+            [Argument(Tooltip = "The ratio used for rescaling of the image before doing actual OCR. Default is 2.0. Higher values are better for recognizing details like small fonts.")]
```
[the remark this answers is not in the document]
Comment (Vegebutcher) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: my proposed tooltip for sensitivity:
Factor of image zoom that allows better recognition of smaller text
This and other tooltips are here:
https://github.com/G1ANT-Robot/G1ANT.Addon/pull/86

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 4]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
                 using (var tEngine = new TesseractEngine(dataPath, language, EngineMode.TesseractAndCube))
-                using (var img = Pix.LoadFromFile(imagePath))
+                using (var img = PixConverter.ToPix(imgToParse))
                 using (var page = tEngine.Process(img))
                 {
                     var rectResult = new Rectangle(-1, -1, -1, -1);
-                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0));
-                    if (wordsWithRectPositions.ContainsValue(search))
+                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0), arguments.Sensitivity.Value);
+                    var searchWords = search.Split(' ');
```
Comment (danieladamkaiser) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: Tooltip says "Provide only single words" is it still true? If not change tooltip (ask @Vegebutcher to confirm text if necessary)

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs [hunk 5]
```diff
@@ -21,37 +22,46 @@ public class OcrOfflineFromScreenCommand : Command
         public class Arguments : CommandArguments
         {
             [Argument(Required = true)]
-            public RectangleStructure Area { get; set; }
+            public RectangleStructure Area { get; set; } = new RectangleStructure(SystemInformation.VirtualScreen);
```
Comment (danieladamkaiser) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs]: Its shouldnt be required with default value

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs [hunk 6]
```diff
@@ -21,37 +22,46 @@ public class OcrOfflineFromScreenCommand : Command
...
+                throw new ArgumentException("Argument Area is not a valid rectangle");
+            var rectangle = arguments.Area.Value;
+            if (arguments.Relative.Value)
+            {
+                var foregroundWindowRect = new RobotWin32.Rect();
+                RobotWin32.GetWindowRectangle(RobotWin32.GetForegroundWindow(), ref foregroundWindowRect);
+                rectangle = new Rectangle(rectangle.X + foregroundWindowRect.Left,
+                    rectangle.Y + foregroundWindowRect.Top,
+                    foregroundWindowRect.Right - foregroundWindowRect.Left,
+                    foregroundWindowRect.Bottom - foregroundWindowRect.Top);
```
Comment (danieladamkaiser) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs]: Same as up - it looks like you are getting wrong width/heigth

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vegebutcher) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: no, it's no longer true after this refactor

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs (hunk 1 above)
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: Not going to support negative width and height.

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 7]
```diff
@@ -78,13 +96,18 @@ public void Execute(Arguments arguments)
...
+        private Rectangle UnionRectangles(List<Rectangle> rectangleList)
+        {
+            int xMin = rectangleList.Min(s => s.X);
+            int yMin = rectangleList.Min(s => s.Y);
+            int xMax = rectangleList.Max(s => s.X + s.Width);
+            int yMax = rectangleList.Max(s => s.Y + s.Height);
+            return new Rectangle(xMin, yMin, xMax - xMin, yMax - yMin);
+        }
+
+        public Dictionary<Rectangle, string> GetWords(string tesseractHtml, double imgRescaleRatio)
```
[the remark this answers is not in the document]
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: Rectangle is a key and found text value is a value seams legit for me.

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs (hunk 5 above)
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 8]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
                 using (var page = tEngine.Process(img))
                 {
                     var rectResult = new Rectangle(-1, -1, -1, -1);
-                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0));
-                    if (wordsWithRectPositions.ContainsValue(search))
+                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0), arguments.Sensitivity.Value);
+                    var searchWords = search.Split(' ');
+                    if (searchWords.Length > 1)
+                    {
+                        rectResult = UnionRectangles(wordsWithRectPositions.Where(x => searchWords.Contains(x.Value)).Select(a => a.Key).ToList());
```
[the remark this answers is not in the document]
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 9]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
             {
                 using (var tEngine = new TesseractEngine(dataPath, language, EngineMode.TesseractAndCube))
-                using (var img = Pix.LoadFromFile(imagePath))
+                using (var img = PixConverter.ToPix(imgToParse))
                 using (var page = tEngine.Process(img))
                 {
                     var rectResult = new Rectangle(-1, -1, -1, -1);
-                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0));
-                    if (wordsWithRectPositions.ContainsValue(search))
+                    var wordsWithRectPositions = GetWords(page.GetHOCRText(0), arguments.Sensitivity.Value);
```
[the remark this answers is not in the document]
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs (hunk 2 above)
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs [hunk 10]
```diff
@@ -37,31 +38,48 @@ public class Arguments : CommandArguments
...
         {
         }
 
-        double imgRescaleRatio = 4.0;
         public void Execute(Arguments arguments)
         {
-            var rectangle = !arguments.Relative.Value ? arguments.Area.Value : arguments.Area.Value.ToAbsoluteCoordinates();
+            if (!arguments.Area.Value.IsValidRectangle())
+                throw new ArgumentException("Argument Area is not a valid rectangle");
+            var rectangle = arguments.Area.Value;
```
[the remark this answers is not in the document]
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DanielWilczek) [reply G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFindCommand.cs]: done

File: G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs [hunk 11]
```diff
@@ -20,38 +20,47 @@ public class OcrOfflineFromScreenCommand : Command
...
             [Argument]
             public BooleanStructure Relative { get; set; } = new BooleanStructure(true);
 
             [Argument]
             public VariableStructure Result { get; set; } = new VariableStructure("result");
 
-            [Argument(DefaultVariable = "timeoutocr")]
-            public override TimeSpanStructure Timeout { get; set; }
-
             [Argument(Tooltip = "The language which should be considered trying to recognize text")]
```
Comment (Vegebutcher) [inline G1ANT.Addon.OCR.Tesseract/G1ANT.Addon.OCR.Tesseract/Commands/OcrOfflineFromScreenCommand.cs]:
```suggestion
            [Argument(Tooltip = "Language to be used for text recognition")]
```

**Comment by danieladamkaiser** [comment]
You have my approve cause we need this change on rc fast, but you should consider cleaning up Execute method a little as it has grown to enormous size

**Pull request merged**